pkg/decoder/message_producer.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2017 Datadog, Inc.
+
+package decoder
+
+import (
+	"bytes"
+	"errors"
+	"regexp"
+	"sync"
+	"time"
+)
+
+// messageProducer consumes new lines to compute and emit messages
+// Prepare and Dispose can be used for special needs
+type messageProducer interface {
+	Start()
+	Stop()
+	Consume(line []byte)
+}
+
+// singleLineMessageProducer emits new messages whenever it receives new lines
+type singleLineMessageProducer struct {
+	outputChan chan *Message
+	isMsgTrunc bool
+}
+
+// newSingleLineMessageProducer returns a new singleLineMessageProducer
+func newSingleLineMessageProducer(outputChan chan *Message) *singleLineMessageProducer {
+	return &singleLineMessageProducer{
+		outputChan: outputChan,
+	}
+}
+
+// Start is not needed for single line messages
+func (p *singleLineMessageProducer) Start() {
+	// not implemented
+}
+
+// Stop is not needed for single line messages
+func (p *singleLineMessageProducer) Stop() {
+	// not implemented
+}
+
+// Consume checks line and the inner state of the producer to emit new messages
+func (p *singleLineMessageProducer) Consume(line []byte) {
+	lineLen := len(line)
+	if lineLen == 0 {
+		return
+	}
+	var msg *Message
+	if lineLen < maxMessageLen {
+		msg = newMessage(line, p.isMsgTrunc, lineLen+1) // add 1 for `\n`
```

Comment:
I would consider :
```
isMsgTrunc = lineLen > maxMessageLen
```
and / or adding a dedicated constructor `newTruncatedMessage` that adds 1 to the line length and possibly sets the flag to true if needed.